pathology/hovernet/training_ignite.py

```diff
@@ -0,0 +1,303 @@
+    valid_data = [data[i] for i in valid_indices]
+
+    return train_data, valid_data
+
+def cropping_center(x, crop_shape, batch=False):
+    """Crop an input image at the centre.
+    Args:
+        x: input array
+        crop_shape: dimensions of cropped array
+
+    Returns:
+        x: cropped array
+
+    """
+    orig_shape = x.shape
+    if not batch:
+        h0 = int((orig_shape[1] - crop_shape[0]) * 0.5)
+        w0 = int((orig_shape[2] - crop_shape[1]) * 0.5)
+        x = x[:, h0: h0 + crop_shape[0], w0: w0 + crop_shape[1]]
+    else:
+        h0 = int((orig_shape[2] - crop_shape[0]) * 0.5)
+        w0 = int((orig_shape[3] - crop_shape[1]) * 0.5)
+        x = x[..., h0: h0 + crop_shape[0], w0: w0 + crop_shape[1]]
+    return x
+
+def val_post_process(data):
+    seg_pred = data["pred"][HoVerNetBranch.NP.value]
+    trans = Compose([Activations(softmax=True), AsDiscrete(argmax=True, to_onehot=2)])
+    data["pred"][HoVerNetBranch.NP.value] = trans(seg_pred)
+
+    return data
+
+def train_post_process(data):
+    seg_pred = data["pred"][HoVerNetBranch.NP.value]
+    trans = Compose([Activations(softmax=True), AsDiscrete(argmax=True, to_onehot=2)])
+    data["pred"][HoVerNetBranch.NP.value] = trans(seg_pred)
+
+    return data
+
+def _from_engine(keys):
+    keys = ensure_tuple(keys)
+
+    def _wrapper(data):
+        ret = [[i[k][HoVerNetBranch.NP.value] for i in data] for k in keys]
+        return tuple(ret) if len(ret) > 1 else ret[0]
+
+    return _wrapper
+
+
+class PrepareBatchExtraInput_v2():
+    def __init__(self, extra_keys) -> None:
+        self.prepare_batch = PrepareBatchExtraInput(extra_keys)
+
+    def __call__(self, batchdata, device, non_blocking, **kwargs):
+        image, label, extra_label, _ = self.prepare_batch(batchdata, device, non_blocking, **kwargs)
+        all_label = {
+            HoVerNetBranch.NP: label,
+            HoVerNetBranch.NC: extra_label[0],
+            HoVerNetBranch.HV: extra_label[1],
+        }
+
+        return image, all_label
+
+
+def run(cfg):
+    log_dir = create_log_dir(cfg)
+    device = set_device(cfg)
+    # --------------------------------------------------------------------------
+    # Data Loading and Preprocessing
+    # --------------------------------------------------------------------------
+    # __________________________________________________________________________
+    # __________________________________________________________________________
+    # Build MONAI preprocessing
+    train_data, valid_data = prepare_datasets(cfg["root"])
+    train_transforms = Compose(
+        [
+            EnsureChannelFirstd(keys=("image", "label"), channel_dim=-1),
+            SplitDimd(keys="label", output_postfixes=["inst", "type"]),
+            ComputeHoVerMapsd(keys="label_inst"),
+            CastToTyped(keys=["image", "label_inst", "label_type", "hover_label_inst"], dtype=torch.float32),
+            Lambdad(keys="label", func=lambda x: x[1: 2, ...] > 0),
```

Comment:
Is there any MONAI intensity transform can work as this Lambda?
If not, maybe we can enhance the MONAI transform:
https://github.com/Project-MONAI/MONAI/blob/dev/monai/transforms/intensity/dictionary.py#L722

Thanks.

Reply:
Hi @Nic-Ma, thanks for your feedback, since the original label has 2 channels, which represent instance map and type map individually, `ThresholdIntensity` can threshold the intensity correctly, but here I only want one channel to be the binary segmentation mask. I think maybe I need `Lambdad` here and since it is not a common use case, we don't need to enhance the transforms then? Thanks!

Reply:
ok, sounds good.

Thanks.